core/src/main/scala/chisel3/internal/MonoConnect.scala

```diff
@@ -389,9 +389,45 @@ private[chisel3] object MonoConnect {
     _source:             Element,
     context_mod:         RawModule
   ): Unit = {
-    import BindingDirection.{Input, Internal, Output} // Using extensively so import these
+    // Reify sink and source if they're views.
     val sink = reify(_sink)
     val source = reify(_source)
+
+    checkConnect(sourceInfo, sink, source, context_mod)
+    issueConnect(sink, source)
+  }
+}
+
+private[chisel3] object checkConnect {
+  private[chisel3] def apply(
+    implicit sourceInfo: SourceInfo,
+    _sink:               Element,
+    _source:             Element,
+    context_mod:         RawModule
+  ): Unit = {
+    checkConnection(sourceInfo, _sink, _source, context_mod)
+  }
+
+  // This function checks if element-level connection operation allowed.
+  // It throws the appropriate exception, if necessary.
+  private def checkConnection(
+    implicit sourceInfo: SourceInfo,
+    sink:                BaseType,
+    source:              BaseType,
+    context_mod:         RawModule
+  ): Unit = {
```

Comment:
```suggestion
  private def checkConnection(
    sourceInfo: SourceInfo,
    sink:                BaseType,
    source:              BaseType,
    context_mod:         RawModule
  ): Unit = {
```
Use of implicit is an antipattern here.